packages/ui/src/react/swapkit-widget.tsx

```diff
@@ -1,155 +1,205 @@
+    }
+  }, [inputAsset, outputAsset, amount, swapKit]);
 
   useEffect(() => {
-    void loadTokens();
-  }, [loadTokens]);
-
-  return (
-    <SwapKitProvider config={swapKitConfig} plugins={config?.plugins}>
-      <SwapKitContent availableAssets={availableAssets} tokensLoaded={tokensLoaded} />
-    </SwapKitProvider>
-  );
-}
+    void updateEstimatedOutput();
+  }, [updateEstimatedOutput]);
+
+  const handleSwap = async (route: QuoteResponseRoute) => {
+    if (!swapKit) return;
+
+    try {
+      setIsSwapping(true);
+      const swap = await swapKit.swap({ route });
+
+      await swap.wait();
+      setAmount("");
+      setEstimatedOutput(undefined);
+      setRoutes([]);
+      toast.success("Swap completed successfully");
+    } catch (error) {
+      console.error("Swap failed:", error);
+      toast.error(`Swap failed: ${error instanceof Error ? error.message : "Unknown error"}`);
+    } finally {
+      setIsSwapping(false);
+    }
+  };
 
-const initialState = {
-  inputAsset: undefined as AssetValue | undefined,
-  isConnected: false,
-  outputAsset: undefined as AssetValue | undefined,
-};
+  const swap = async (route: QuoteResponseRoute, inputAssetValue?: AssetValue) => {
+    if (!(inputAssetValue && swapKit)) return;
 
-function SwapKitContent({ availableAssets, tokensLoaded }: { availableAssets?: AssetValue[]; tokensLoaded: boolean }) {
-  const [{ isConnected, inputAsset, outputAsset }, dispatch] = useReducer(reducer, initialState);
-  const setInputAsset = useCallback((inputAsset: AssetValue) => {
-    dispatch({ inputAsset, type: "setInputAsset" });
-  }, []);
-
-  const setInputAmount = useCallback(
-    (inputAmount: string) => {
-      if (inputAsset) {
-        inputAsset.set(inputAmount);
+    try {
+      const isChainflip = route?.providers?.includes(ProviderName.CHAINFLIP);
+      if (isChainflip) {
+        await handleSwap(route);
+        return;
       }
-    },
-    [inputAsset],
-  );
-
-  const setOutputAsset = useCallback((outputAsset: AssetValue) => {
-    dispatch({ outputAsset, type: "setOutputAsset" });
-  }, []);
 
-  const setOutputAmount = useCallback(
-    (outputAmount: string) => {
-      if (outputAsset) {
-        outputAsset.set(outputAmount);
+      const tx = route.tx;
+      if (!tx || typeof tx === "string" || !("from" in tx)) {
+        throw new Error("Invalid transaction format");
       }
-    },
-    [outputAsset],
-  );
 
-  if (!tokensLoaded) {
-    return <div>Loading...</div>;
-  }
+      const isApproved = await swapKit.isAssetValueApproved(inputAssetValue, tx.from);
+      if (isApproved) {
+        await handleSwap(route);
+      } else {
+        await swapKit.approveAssetValue(inputAssetValue, tx.from);
+        toast.success("Asset approved, you can now swap");
+      }
+    } catch (error) {
+      console.error("Swap process failed:", error);
+      toast.error(`Swap process failed: ${error instanceof Error ? error.message : "Unknown error"}`);
+    }
+  };
```

Comment:
I don't get the whole picture of this logic but it seems `overengineered` there.

Most of those handlers are done via swapKit.swap afaik and when it comes to `approved` assets - they have to be checked for all of the providers imo 🤔 gas assets will return true anyway, so can call that without an issue.